### PR TITLE
[Amazon-AMC] | Added regex validation for externalUserId in Amazon AMC

### PR DESCRIPTION
**Pull request by Innovative-GauravKochar**
This pull request adds regex validation for the externalUserId in Amazon AMC to prevent failures in the destination API.
Tested [the fix in staging](https://docs.google.com/document/d/1rHTMYd-7GGv5dH0FSxMJVa4TZBlOohQltVxJRf9yXB8/edit?tab=t.0#heading=h.x9dgiyfuowsl), and it's working fine!

Jira ticket:- https://segment.atlassian.net/jira/software/c/projects/STRATCONN/boards/310?assignee=63617339fc0cc7a600b03c6b&selectedIssue=STRATCONN-5293

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [x] [If destination is already live] Tested for backward compatibility of destination. **Note:** New required fields are a breaking change.
- [x] [Segmenters] Tested in the staging environment
- [ ] [Segmenters] [If applicable for this change] Tested for regression with Hadron.

### CODE REVIEW

File: packages/destination-actions/src/destinations/amazon-amc/syncAudiencesToDSP/index.ts [hunk 1]
```diff
@@ -125,6 +126,11 @@ async function processPayload(
 ) {
   const payloadRecord = createPayloadToUploadRecords(payload, audienceSettings)
   // Regular expression to find a audienceId numeric string and replace the quoted audienceId string with an unquoted number
+  if (!payloadRecord.records?.length) {
+    throw new PayloadValidationError(
+      'externalUserId must satisfy regular expression pattern: [0-9a-zA-Z\\-\\_]{1,128}"}'
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/amazon-amc/syncAudiencesToDSP/index.ts:131]: Is there a typo in this regex?

File: packages/destination-actions/src/destinations/amazon-amc/syncAudiencesToDSP/index.ts [hunk 2]
```diff
@@ -145,6 +151,11 @@ function createPayloadToUploadRecords(payloads: Payload[], audienceSettings: Aud
   const records: AudienceRecord[] = []
   const { audienceId } = payloads[0]
   payloads.forEach((payload: Payload) => {
+    // Check if the externalUserId matches the pattern
+    if (!REGEX_EXTERNALUSERID.test(payload.externalUserId)) {
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/amazon-amc/syncAudiencesToDSP/index.ts:155]: This destination might need multistatus. We should not silently drop events 😅

Comment (Innovative-GauravKochar) [reply packages/destination-actions/src/destinations/amazon-amc/syncAudiencesToDSP/index.ts:155]: Yes, we can add multi-status to this destination. I've created a separate [ticket](https://segment.atlassian.net/browse/STRATCONN-5314) in the backlog for it.

File: packages/destination-actions/src/destinations/amazon-amc/syncAudiencesToDSP/index.ts [hunk 3]
```diff
@@ -125,6 +126,11 @@ async function processPayload(
 ) {
   const payloadRecord = createPayloadToUploadRecords(payload, audienceSettings)
   // Regular expression to find a audienceId numeric string and replace the quoted audienceId string with an unquoted number
+  if (!payloadRecord.records?.length) {
+    throw new PayloadValidationError(
+      'externalUserId must satisfy regular expression pattern: [0-9a-zA-Z\\-\\_]{1,128}}'
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/amazon-amc/syncAudiencesToDSP/index.ts:131]: Why is this code throwing a regex error when there are no records?

Comment (varadarajan-tw) [reply packages/destination-actions/src/destinations/amazon-amc/syncAudiencesToDSP/index.ts:131]: Could we move this into `createPayloadToUploadRecords`?

**Review by varadarajan-tw — APPROVED** [review]
Thanks for addressing the comments! ❇️

**Pull request merged**